src/Lean/Meta/ProofBelow.lean

```diff
@@ -0,0 +1,373 @@
+
+  mkIndValConst (indVal : InductiveVal) : Expr :=
+    mkConst indVal.name $ indVal.levelParams.map mkLevelParam
+    
+partial def mkCtorType 
+    (ctx : Context) 
+    (proofBelowIdx : Nat) 
+    (originalCtor : ConstructorVal) : MetaM Expr := 
+  forallTelescope originalCtor.type fun xs t => addHeaderVars 
+    { innerType := t
+      indVal := #[]
+      motives := #[]
+      params := xs[:ctx.numParams]
+      args := xs[ctx.numParams:]
+      target := xs[:ctx.numParams] }
+where
+  addHeaderVars (vars : Variables) := do
+    let headersWithNames ← ctx.headers.mapIdxM fun idx header =>
+      (ctx.proofBelowNames[idx], fun _ => pure header)
+
+    withLocalDeclsD headersWithNames fun xs =>
+      addMotives { vars with indVal := xs } 
+
+  addMotives (vars : Variables) := do
+    let motiveBuilders ← ctx.motives.mapM fun (motiveName, motiveType) => 
+      (motiveName, BinderInfo.implicit, fun _ => 
+        instantiateForall motiveType vars.params)
+    withLocalDecls motiveBuilders fun xs => 
+      modifyBinders { vars with target := vars.target ++ xs, motives := xs } 0
+
+  modifyBinders (vars : Variables) (i : Nat) := do
+    if i < vars.args.size then
+      let binder := vars.args[i]
+      let binderType := ←inferType binder
+      if ←checkCount binderType then
+        mkProofBelowBinder vars binder binderType fun indValIdx x =>
+          mkMotiveBinder vars indValIdx binder binderType fun y =>
+            modifyBinders { vars with target := vars.target ++ #[binder, x, y]} i.succ
+      else modifyBinders { vars with target := vars.target.push binder } i.succ
+    else rebuild vars
+
+  rebuild (vars : Variables) := 
+    vars.innerType.withApp fun f args => do
+      let hApp := 
+        mkAppN 
+          (mkConst originalCtor.name $ ctx.typeInfos[0].levelParams.map mkLevelParam)
+          (vars.params ++ vars.args)
+      let innerType := mkAppN vars.indVal[proofBelowIdx] $
+        vars.params ++ vars.motives ++ args[ctx.numParams:] ++ #[hApp]
+      let x ← mkForallFVars vars.target innerType 
+      replaceTempVars vars x
+
+  replaceTempVars (vars : Variables) (ctor : Expr) :=
+    let levelParams := 
+      ctx.typeInfos[0].levelParams.map mkLevelParam
+
+    ctor.replaceFVars vars.indVal $ ctx.proofBelowNames.map fun indVal =>
+      mkConst indVal levelParams
+  
+  checkCount (domain : Expr) : MetaM Bool := do
+    let run (x : StateRefT Nat MetaM Expr) : MetaM (Expr × Nat) := StateRefT'.run x 0
+    let (_, cnt) := ←run $ transform domain fun e => do
+      if let some name := e.constName? then
+        if let some idx := ctx.typeInfos.findIdx? fun indVal => indVal.name == name then
+          let cnt := ←get
+          set $ cnt + 1
+      TransformStep.visit e
+
+    if cnt > 1 then 
+      let message := 
+        "only arrows are allowed as premises. Multiple recursive occurrences detected:\n  "
+        ++ (←ppExpr domain)
+      throwError message
```

Comment:
We don't eagerly convert expressions into format objects. The motivation is not performance, but flexibility and the ability to traverse error messages, and analyzing them. The `indentExpr` function is also useful to make sure our error messages are uniform.
```lean
throwError "only arrows are allowed as premises. Multiple recursive occurrences detected:{indentExpr domain}"
```

Reply:
done.